TestControllers/TestController.psm1

```diff
@@ -748,4 +764,95 @@ Class TestController
 			Write-LogErr "Source: Line $line in script $script_name."
 		}
 	}
+
+	[bool] GetAndCompareOsLogs($AllVMData, $Status) {
+		$retValue = $true
+		try	{
+			if (!($status -imatch "Initial" -or $status -imatch "Final")) {
+				Write-LogInfo "Status value should be either final or initial"
+				return $false
+			}
+			foreach ($VM in $AllVMData) {
+				Write-LogInfo "Collecting $($VM.RoleName) VM Kernel $status Logs ..."
+
+				$bootLogDir = "$global:Logdir\$($VM.RoleName)"
+				mkdir $bootLogDir -Force | Out-Null
+
+				$currentBootLogFile = "${status}BootLogs.txt"
+				$currentBootLog = Join-Path $BootLogDir $currentBootLogFile
+
+				$initialBootLogFile = "InitialBootLogs.txt"
+				$initialBootLog = Join-Path $BootLogDir $initialBootLogFile
+
+				$kernelLogStatus = Join-Path $BootLogDir "KernelLogStatus.txt"
+
+				if ($this.EnableCodeCoverage -and ($status -imatch "Final")) {
+					Write-LogInfo "Collecting coverage debug files from VM $($VM.RoleName)"
+
+					$gcovCollected = Collect-GcovData -ip $VM.PublicIP -port $VM.SSHPort `
+						-username $global:user -password $global:password -logDir $global:LogDir
+
+					if ($gcovCollected) {
+						Write-LogInfo "GCOV data collected successfully"
+					} else {
+						Write-LogErr "Failed to collect GCOV data from VM: $($VM.RoleName)"
+					}
+				}
+
+				Run-LinuxCmd -ip $VM.PublicIP -port $VM.SSHPort -runAsSudo `
+					-username $global:user -password $global:password `
+					-command "dmesg > ./${currentBootLogFile}" | Out-Null
+				Copy-RemoteFiles -download -downloadFrom $VM.PublicIP -port $VM.SSHPort -files "./${currentBootLogFile}" `
+					-downloadTo $BootLogDir -username $global:user -password $global:password | Out-Null
+				Write-LogInfo "$($VM.RoleName): $status Kernel logs collected successfully to ${currentBootLogFile} file."
+
+				Write-LogInfo "Checking for call traces in kernel logs.."
+				$KernelLogs = Get-Content $currentBootLog
+				$callTraceFound  = $false
+				foreach ($line in $KernelLogs) {
```

Comment:
Suggestion: This foreach blocks are everywhere in the source code. Can you make it function and call? Code duplication.

Reply:
I suggest we file an issue for this, and file a separate PR